### PR TITLE
🔀✨ nwk create form

### PR DESCRIPTION
**Pull request by IamPekka058**
**Description:**  

Implemented NwkCreateDialog.

**Reference:**

Issue: #16 

Notifying additional team members:

@mirrodi @MrSebastian @AnHo314

---
**Definition of Done (DoD):**

- [x] Unittests gepflegt
- [x] Dokumentation ist gepflegt

### CODE REVIEW

**Comment by IamPekka058** [comment]
Die package-lock.json habe ich nicht hochgeladen, da diese als Quelle das artifactory.muenchen.de verwendet. (Für OpenSource falsch)

File: praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue [hunk 1]
```diff
@@ -31,6 +25,8 @@
 import ExcelImportNwk from "@/components/nachwuchskraefte/ExcelImportNwk.vue";
 import PageTitle from "@/components/common/PageTitle.vue";
 import ActiveNwkList from "@/components/nachwuchskraefte/ActiveNwkList.vue";
+import router from "@/router";
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue:28]: unused import should be removed

**Review by Tuckerle — CHANGES_REQUESTED** [review]

File: praktikumsplaner-frontend/frontend/src/types/NwkCreate.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+        public vorlesungstage: string[],
+        public studiengang?: string,
+        public ausbildungsrichtung?: string
+    ) {}
+}
+
+// Used in NwkCards to determine if the details button should be shown
+export function hasDetails(nwk: NwkCreate): boolean {
+    return nwk.vorlesungstage.length > 0;
+}
```
[the remark this answers is not in the document]
Comment (IamPekka058) [reply praktikumsplaner-frontend/frontend/src/types/NwkCreate.ts]: [f1f229f](https://github.com/it-at-m/Praktikumsplaner/pull/214/commits/f1f229f6d337edf761bd8f0b7e4ab252d3a1e34b) Fixed this

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                    <v-list>
+                        <v-row>
+                            <v-col>
+                                <v-list-item>
+                                    <v-text-field
+                                        v-model="nwk.vorname"
+                                        label="Vorname"
+                                        :rules="nameRule"
+                                        outlined
+                                    ></v-text-field>
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:31]: Could/Should be extracted into separate component.

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                                </v-list-item>
+                            </v-col>
+                            <v-col>
+                                <v-list-item>
+                                    <v-text-field
+                                        v-model="nwk.nachname"
+                                        label="Nachname"
+                                        :rules="nameRule"
+                                        outlined
+                                    ></v-text-field>
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:41]: Could/Should be extracted into separate component.

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                        </v-row>
+                        <v-row>
+                            <v-col>
+                                <v-list-item>
+                                    <v-text-field
+                                        v-model="nwk.jahrgang"
+                                        label="Jahrgang"
+                                        :rules="jahrgangRule"
+                                        outlined
+                                    ></v-text-field>
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:53]: Could/Should be extracted into separate component.

**Review by Tuckerle — CHANGES_REQUESTED** [review]

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:1]: This is not documented in docs

Comment (IamPekka058) [reply praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:1]: Did https://github.com/it-at-m/Praktikumsplaner/pull/214/commits/3b16c979f23c1cbf06bfb3da7ea314abd69ef1c8 fix that?

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                                        item-value="value"
+                                        item-text="name"
+                                        outlined
+                                        clearable
+                                        :rules="isStudiumOrAusbildungRule"
+                                        @click:clear="
+                                            nwk.ausbildungsrichtung = undefined
+                                        "
+                                    ></v-select>
+                                </v-list-item>
```
[the remark this answers is not in the document]
Comment (IamPekka058) [reply praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue]: Resolved by https://github.com/it-at-m/Praktikumsplaner/pull/214/commits/411ff015cffde28e3242fcf2677a0dd333358dfd ?

File: praktikumsplaner-frontend/frontend/src/components/common/JahrgangInput.vue [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+const jahrgangRule = [
+    validationRules.notEmptyRule("Der Jahrgang darf nicht leer sein."),
+    validationRules.regexRule(
+        /^([0-9]{2})\/([0-9]{2})$/,
+        "Der Jahrgang muss im Format XX/XX angegeben werden."
+    ),
+];
+
+const props = defineProps<{
+    nwk: NwkCreate;
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/common/JahrgangInput.vue:25]: This should be implemented in a way that it supports the v-model directive

File: praktikumsplaner-frontend/frontend/src/components/common/NameInput.vue
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/common/NameInput.vue:1]: I think the v-if structure is not the best design. I would either split it up into two different components or use this component for both at the same time so it always shows Vorname and Nachname similar to StudienOrAusbildungsRichtungSelect. Another option would be to generalize the input so all the things that differ between vorname and nachname can be adjusted when creating.

File: praktikumsplaner-frontend/frontend/src/components/common/NameInput.vue [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    ),
+    validationRules.maxLengthRule(
+        255,
+        "Der Name darf maximal 255 Zeichen lang sein."
+    ),
+    validationRules.notEmptyRule("Der Name darf nicht leer sein."),
+];
+
+const props = defineProps<{
+    nwk: NwkCreate;
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/common/NameInput.vue:37]: This should be implemented in a way that it supports the v-model directive

File: praktikumsplaner-frontend/frontend/src/components/common/StudienrichtungOrAusbildungsrichtungSelect.vue [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+</template>
+
+<script setup lang="ts">
+import { Studiengang } from "@/types/Studiengang";
+import { Ausbildungsrichtung } from "@/types/Ausbildungsrichtung";
+import NwkCreate from "@/types/NwkCreate";
+import { computed } from "vue";
+
+const props = defineProps<{
+    nwk: NwkCreate;
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/common/StudienrichtungOrAusbildungsrichtungSelect.vue:45]: This should be implemented in a way that it supports the v-model directive

File: praktikumsplaner-frontend/frontend/src/components/common/StudienrichtungOrAusbildungsrichtungSelect.vue [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                    <v-select
+                        v-model="nwkProp.ausbildungsrichtung"
+                        label="Ausbildungsrichtung"
+                        :items="Ausbildungsrichtung"
+                        item-value="value"
+                        item-text="name"
+                        outlined
+                        clearable
+                        :rules="isStudiumOrAusbildungRule"
+                        @click:clear="nwkProp.ausbildungsrichtung = undefined"
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/components/common/StudienrichtungOrAusbildungsrichtungSelect.vue:30]: Even if its just one line of code, it should be put into a method so there is no/less logic in the template. Additionally if you need to do anything else when clearing for example showing a message or sending a request or trigger an event the code in the template would become hard to read and maintain.

**Review by Tuckerle — CHANGES_REQUESTED** [review]

File: doc/src/features/Nachwuchskraefte.md [hunk 11]
```diff
@@ -42,6 +42,21 @@ The days can be chained together by plus signs, for example: Mo + Di + Mi.
...
+### Manually creating [Nwks](./../glossary.md#nwk)
+In the top right of the site is a blue upload button labeled `Erstellen`.
+
+It is possible to create a [Nwks](./../glossary.md#nwk) manually there.
+
+The first name (Vorname) has to be between 2 and 255 characters.
+
+The last name (Nachname) has to be between 2 and 255 characters.
+
+There a two selectors to select either the training direction or course of study
```
Comment (LuisEinig) [inline doc/src/features/Nachwuchskraefte.md:54]:
```suggestion
There are two selectors to select either the training direction or course of study
```
typo

**Comment by LuisEinig** [comment]
grafik
grafik

When selecting both training and study the fields change position.
The dialog also has a scrollbar for minimal usage, this could be mitigated by making the dialog wider

File: praktikumsplaner-frontend/frontend/src/api/NwkService.ts [hunk 12]
```diff
@@ -3,8 +3,18 @@ import { API_BASE, NWK_BASE } from "@/constants";
...
 import { useSnackbarStore } from "@/stores/snackbar";
 import { Levels } from "@/api/Error";
+import NwkCreate from "@/types/NwkCreate";
 
 export default {
+    saveNwk(nwk: NwkCreate): Promise<void> {
+        return fetch(`${API_BASE}${NWK_BASE}`, FetchUtils.getPOSTConfig(nwk))
+            .then((response) => {
+                FetchUtils.defaultResponseHandler(response);
+            })
```
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/api/NwkService.ts:13]: There is no Feedback to the user if the creation was successful or not.

**Review by MrSebastian — CHANGES_REQUESTED** [review]

File: praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue [hunk 13]
```diff
@@ -22,16 +22,15 @@
...
 import NwkCreateDialog from "@/components/nachwuchskraefte/NwkCreateDialog.vue";
+import { EventBus } from "@/stores/event-bus";
 
-const nwkList = ref();
-function created() {
-    nwkList.value.loadAllActiveNwks();
-}
+EventBus.$on("nwkCreated", () => {
+    EventBus.$emit("nwkCreated");
+});
```
Comment (Tuckerle) [inline praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue:33]: That serves no purpose as the eventbus is global. So basically you are emitting the same event twice

**Review by Tuckerle — APPROVED** [review]

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/ActiveNwkList.vue [hunk 14]
```diff
@@ -70,6 +70,10 @@ function getSubtitle(nwk: Nwk): string {
 EventBus.$on("nwkCreated", () => {
     loadAllActiveNwks();
 });
+
+defineExpose({
+    loadAllActiveNwks,
+});
```
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/ActiveNwkList.vue:76]: is this still needed with the use of EventBus?

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+function saveNwk() {
+    if (!form.value?.validate()) {
+        return;
+    }
+    loading.value = true;
+    cancel();
+    NwkService.saveNwk(nwk.value).finally(() => {
+        loading.value = false;
+        EventBus.$emit("nwkCreated");
+    });
```
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/NwkCreateDialog.vue:110]: the emit should be moved into a `then` bc if an error occurs the event should not be called

File: praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/VorlesungstageSelect.vue [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+const nwk = computed({
+    // getter
+    get() {
+        return props.value;
+    },
+    // setter
+    set(newValue) {
+        emits("input", newValue);
+    },
+});
```
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/components/nachwuchskraefte/VorlesungstageSelect.vue:46]: style is not consistent with other components like e.g. NameInput

File: praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue [hunk 17]
```diff
@@ -10,19 +10,13 @@
...
-                    Hinzufügen
-                </v-btn>
+                <NwkCreateDialog></NwkCreateDialog>
             </v-col>
         </v-row>
         <v-row></v-row>
         <v-container class="box">
             <span> Übersicht</span>
-            <active-nwk-list></active-nwk-list>
+            <active-nwk-list ref="nwkList"></active-nwk-list>
```
Comment (LuisEinig) [inline praktikumsplaner-frontend/frontend/src/views/nachwuchskraefte/NachwuchskraefteView.vue:19]: is this ref still needed?

**Review by MrSebastian — APPROVED** [review]

**Review by LuisEinig — APPROVED** [review]

**Pull request merged**